exp/controllers/awsmachinepool_controller.go

```diff
@@ -496,29 +519,56 @@ func (r *AWSMachinePoolReconciler) findASG(machinePoolScope *scope.MachinePoolSc
+				machinePoolScope.Info("DesiredCapacity diff detected", "incoming", *machinePoolScope.MachinePool.Spec.Replicas, "existing", *existingASG.DesiredCapacity)
 				return true
 			}
 		} else if existingASG.DesiredCapacity != nil {
+			machinePoolScope.Info("Replicas nil, DesiredCapacity diff detected", "incoming", machinePoolScope.MachinePool.Spec.Replicas, "existing", *existingASG.DesiredCapacity)
 			return true
 		}
 	}
 
 	if machinePoolScope.AWSMachinePool.Spec.MaxSize != existingASG.MaxSize {
+		machinePoolScope.Info("MinSize diff detected", "incoming", machinePoolScope.AWSMachinePool.Spec.MaxSize, "existing", existingASG.MaxSize)
 		return true
 	}
 
 	if machinePoolScope.AWSMachinePool.Spec.MinSize != existingASG.MinSize {
+		machinePoolScope.Info("MinSize diff detected", "incoming", machinePoolScope.AWSMachinePool.Spec.MinSize, "existing", existingASG.MinSize)
 		return true
 	}
 
 	if machinePoolScope.AWSMachinePool.Spec.CapacityRebalance != existingASG.CapacityRebalance {
+		machinePoolScope.Info("CapacityRebalance diff detected", "incoming", machinePoolScope.AWSMachinePool.Spec.CapacityRebalance, "existing", existingASG.CapacityRebalance)
 		return true
 	}
 
+	// can't assume a diff here means the config has drifted, optional fields may be nil and be detected falsely as different
 	if !cmp.Equal(machinePoolScope.AWSMachinePool.Spec.MixedInstancesPolicy, existingASG.MixedInstancesPolicy) {
-		machinePoolScope.Info("got a mixed diff here", "incoming", machinePoolScope.AWSMachinePool.Spec.MixedInstancesPolicy, "existing", existingASG.MixedInstancesPolicy)
-		return true
+		mixedInstancesPolicy := machinePoolScope.AWSMachinePool.Spec.MixedInstancesPolicy
+
+		// if both are set, examine the optional fields first
+		if mixedInstancesPolicy != nil && existingASG.MixedInstancesPolicy != nil {
+			// do not detect a diff if we did not set the optional field InstancesDistribution
+			if mixedInstancesPolicy.InstancesDistribution != nil {
+				if !cmp.Equal(mixedInstancesPolicy.InstancesDistribution, existingASG.MixedInstancesPolicy.InstancesDistribution) {
+					machinePoolScope.Info("MixedInstancesPolicy.InstancesDistribution diff detected", "incoming", spew.Sprintf("%#v", mixedInstancesPolicy.InstancesDistribution), "existing", spew.Sprintf("%#v", existingASG.MixedInstancesPolicy.InstancesDistribution))
+					return true
+				}
+			}
+			if !cmp.Equal(mixedInstancesPolicy.Overrides, existingASG.MixedInstancesPolicy.Overrides) {
+				machinePoolScope.Info("MixedInstancesPolicy.Overrides diff detected", "incoming", spew.Sprintf("%#v", mixedInstancesPolicy.Overrides), "existing", spew.Sprintf("%#v", existingASG.MixedInstancesPolicy.Overrides))
+				return true
+			}
+		} else {
+			machinePoolScope.Info("MixedInstancesPolicy diff detected", "incoming", spew.Sprintf("%#v", mixedInstancesPolicy), "existing", spew.Sprintf("%#v", existingASG.MixedInstancesPolicy))
+			return true
+		}
 	}
```

Comment:
I had trouble understanding this.

I think what we want is to compare desired and actual MixedInstancesPolicy, but ignore differences in InstancesDistribution, if the desired InstancesDistribution is nil.

I took a different approach here: https://github.com/dlipovetsky/cluster-api-provider-aws/commit/474881358c32153af3ac8cce710d5e449204c784

What do you think?